postgis/lwgeom_dumpsegments.c

```diff
@@ -0,0 +1,260 @@
+};
+
+/* 32 is the max depth for st_dump, so it seems reasonable
+ * to use the same here
+ */
+#define MAXDEPTH 32
+struct dumpstate {
+	LWGEOM *root;
+	int stacklen; /* collections/geoms on stack */
+	int pathlen;  /* polygon rings and such need extra path info */
+	struct dumpnode stack[MAXDEPTH];
+	Datum path[34]; /* two more than max depth, for ring and point */
+
+	/* used to cache the type attributes for integer arrays */
+	int16 typlen;
+	bool byval;
+	char align;
+
+	uint32_t ring; /* ring of top polygon */
+	uint32_t pt;   /* point of top geom or current ring */
+};
+
+PG_FUNCTION_INFO_V1(LWGEOM_dumpsegments);
+Datum LWGEOM_dumpsegments(PG_FUNCTION_ARGS)
+{
+	FuncCallContext *funcctx;
+	MemoryContext oldcontext, newcontext;
+
+	GSERIALIZED *pglwgeom;
+	LWCOLLECTION *lwcoll;
+	LWGEOM *lwgeom;
+	struct dumpstate *state;
+	struct dumpnode *node;
+
+	HeapTuple tuple;
+	Datum pathpt[2];         /* used to construct the composite return value */
+	bool isnull[2] = {0, 0}; /* needed to say neither value is null */
+	Datum result;            /* the actual composite return value */
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		funcctx = SRF_FIRSTCALL_INIT();
+
+		newcontext = funcctx->multi_call_memory_ctx;
+		oldcontext = MemoryContextSwitchTo(newcontext);
+
+		pglwgeom = PG_GETARG_GSERIALIZED_P_COPY(0);
+		lwgeom = lwgeom_from_gserialized(pglwgeom);
+
+		/* return early if nothing to do */
+		if (!lwgeom || lwgeom_is_empty(lwgeom))
+		{
+			MemoryContextSwitchTo(oldcontext);
+			funcctx = SRF_PERCALL_SETUP();
+			SRF_RETURN_DONE(funcctx);
+		}
+
+		/* Create function state */
+		state = lwalloc(sizeof *state);
+		state->root = lwgeom;
+		state->stacklen = 0;
+		state->pathlen = 0;
+		state->pt = 0;
+		state->ring = 0;
+
+		funcctx->user_fctx = state;
+
+		/*
+		 * Push a struct dumpnode on the state stack
+		 */
+		state->stack[0].idx = 0;
+		state->stack[0].geom = lwgeom;
+		state->stacklen++;
+
+		/*
+		 * get tuple description for return type
+		 */
+		if (get_call_result_type(fcinfo, 0, &funcctx->tuple_desc) != TYPEFUNC_COMPOSITE)
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context that cannot accept a set")));
+		}
+
+		BlessTupleDesc(funcctx->tuple_desc);
+
+		/* get and cache data for constructing int4 arrays */
+		get_typlenbyvalalign(INT4OID, &state->typlen, &state->byval, &state->align);
+
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	/* stuff done on every call of the function */
+	funcctx = SRF_PERCALL_SETUP();
+	newcontext = funcctx->multi_call_memory_ctx;
+
+	/* get state */
+	state = funcctx->user_fctx;
+
+	while (1)
+	{
+		POINTARRAY *points = NULL;
+		LWLINE *line;
+		LWPOLY *poly;
+		POINT4D pt_start, pt_end;
+		POINTARRAY *segment_pa;
+		LWLINE *segment;
+
+		node = &state->stack[state->stacklen - 1];
+		lwgeom = node->geom;
+
+		if (lwgeom->type == LINETYPE || lwgeom->type == POLYGONTYPE)
+		{
+			if (lwgeom->type == LINETYPE)
```

Comment:
Original had triangle too. DumpSegments is also used around constrained Delaunay triangulations, and that means that triangles/TINs can be the thing you try to dump as segments too.